src/validator.h

```diff
@@ -0,0 +1,327 @@
+/* validator.h */
+/*
+ * Copyright information and license terms for this software can be
+ * found in the file LICENSE that is included with the distribution
+ */
+#ifndef VALIDATOR_H
+#define VALIDATOR_H
+
+#include <string>
+#include <set>
+
+#include <epicsThread.h>
+
+#include <pv/pvIntrospect.h>
+
+namespace epics { namespace nt {
+
+/**
+ * @brief Validation methods for NT types.
+ *
+ * @author bsm
+ */
+
+struct Result {
+    struct Error {
+        std::string path;
+        enum Type {
+            MissingField,
+            IncorrectType,
+            IncorrectId,
+        } type;
+
+        Error(std::string const & path, Type type)
+        : path(path), type(type) {}
+
+        bool operator==(const Error& other) const {
+            return type == other.type && path == other.path;
+        }
+
+        std::ostream& dump(std::ostream& os) const {
+            os << "Error(path=" << (path.empty() ? "<root>" : path) << ": ";
+
+            switch(type) {
+                case MissingField:  os << "missing";        break;
+                case IncorrectType: os << "incorrect type"; break;
+                case IncorrectId:   os << "incorrect ID";   break;
+            }
+            os << ")";
+            return os;
+        }
+    };
+
+    epics::pvData::FieldConstPtr field;
+    std::string path;
+    std::vector<Error> errors;
+
+    enum result_t {
+        Pass,
+        Fail,
+    } result;
+
+    Result(const epics::pvData::FieldConstPtr& field, const std::string& path = std::string())
+    : field(field), path(path), errors(), result(Pass) {}
+
+    Result() {}
+
+    Result& operator|=(const Result& other) {
+        result = std::max(result, other.result);
+        errors.insert(errors.end(), other.errors.begin(), other.errors.end());
+        return *this;
+    }
+
+    /**
+     * Clears the contents of this Result.
+     *
+     * @return a reference to itself.
+     */
+    Result& reset(const epics::pvData::FieldConstPtr& newField) {
+        field = newField;
+        path.clear();
+        errors.clear();
+        result = Pass;
+        return *this;
+    }
+
+    /**
+     * Retrieves a Result from a thread-local cache. Allocates a new Result
+     * if it wasn't allocated yet.
+     *
+     * @return a reference to the stored Result.
+     *
+     */
+    static Result& fromCache(epicsThreadOnceId *onceId, epicsThreadPrivateId *cachedId) {
+        epicsThreadOnce(onceId, &once, static_cast<void*>(cachedId));
+
+        Result *result = static_cast<Result*>(epicsThreadPrivateGet(*cachedId));
+
+        if (!result) {
+            result = new Result;
+            epicsThreadPrivateSet(*cachedId, result);
```

Comment:
This looks like a memory leak if called from eg. a PVA callback (on a PVA worker thread) as the Result is not delete'd when the thread ends (eg. IOC restart).

FYI. there is `epicsAtThreadExit()` for this kind of cleanup.  Though this only works for epicsThread s, which is an assumption I'd rather not make.

Also, having to call `epicsThreadOnce()` each time, with it's global mutex, likely negates any performance benefit of a per-thread cache.